src/utils/span32.cairo

```diff
@@ -0,0 +1,176 @@
+    }
+}
+
+impl Span32Index<T> of IndexView<Span32<T>, usize, @T> {
+    #[inline(always)]
+    fn index(self: @Span32<T>, index: usize) -> @T {
+        self.snapshot.index(index)
+    }
+}
+
+trait Array32Trait<T> {
+    fn span(self: @Array<T>) -> Span32<T>;
+}
+
+impl Array32<T> of Array32Trait<T> {
+    #[inline(always)]
+    fn span(self: @Array<T>) -> Span32<T> {
+        assert(self.len() <= 32, 'array too big');
+        Span32 { snapshot: Span { snapshot: self } }
+    }
+}
+
+impl StoreContractAddressSpan of Store<Span32<ContractAddress>> {
+    fn read(
+        address_domain: u32, base: StorageBaseAddress
+    ) -> SyscallResult<Span32<ContractAddress>> {
+        StoreContractAddressSpan::read_at_offset(address_domain, base, 0)
+    }
+
+    fn write(
+        address_domain: u32, base: StorageBaseAddress, value: Span32<ContractAddress>
+    ) -> SyscallResult<()> {
+        StoreContractAddressSpan::write_at_offset(address_domain, base, 0, value)
+    }
+
+    fn read_at_offset(
+        address_domain: u32, base: StorageBaseAddress, mut offset: u8
+    ) -> SyscallResult<Span32<ContractAddress>> {
+        let mut arr: Array<ContractAddress> = ArrayTrait::new();
+
+        // Read the stored array's length. If the length is superior to 255, the read will fail.
+        let len: u8 = Store::<u8>::read_at_offset(address_domain, base, offset)
+            .expect('Storage Span too large');
+        offset += 1;
+
+        // Sequentially read all stored elements and append them to the array.
+        let exit = len + offset;
+        loop {
+            if offset >= exit {
+                break;
+            }
+
+            let value = Store::<ContractAddress>::read_at_offset(address_domain, base, offset)
+                .unwrap();
+            arr.append(value);
+            offset += Store::<ContractAddress>::size();
+        };
+
+        // Return the array.
+        Result::Ok(Array32Trait::span(@arr))
+    }
+
+    fn write_at_offset(
+        address_domain: u32,
+        base: StorageBaseAddress,
+        mut offset: u8,
+        mut value: Span32<ContractAddress>
+    ) -> SyscallResult<()> {
+        // // Store the length of the array in the first storage slot.
+        let len: u8 = value.len().try_into().expect('Storage - Span too large');
+        Store::<u8>::write_at_offset(address_domain, base, offset, len);
+        offset += 1;
+
+        // Store the array elements sequentially
+        loop {
+            match value.pop_front() {
+                Option::Some(element) => {
+                    Store::<ContractAddress>::write_at_offset(
+                        address_domain, base, offset, *element
+                    );
+                    offset += Store::<felt252>::size();
+                },
+                Option::None(_) => {
+                    break Result::Ok(());
+                }
+            };
+        }
+    }
+
+    fn size() -> u8 {
+        255 * Store::<felt252>::size()
```

Comment:
the goal of Span32 is to have a size of 32 max :)

Reply:
oops